### PR TITLE
feat: Differential testing: Enhance contract bytecode dumper to handle modular representation

### PR DESCRIPTION
**Pull request by vtronkov**
**Description**:
* Adding dumpers for contract bytecodes

**Related issue(s)**:

Fixes https://github.com/hashgraph/hedera-services/issues/10994

**Notes for reviewer**:
* Branched off https://github.com/hashgraph/hedera-services/pull/11489
* Added some files from https://github.com/hashgraph/hedera-services/pull/11385 that will be deleted when it's merged(I didn't want to merge two branches into this one)

### CODE REVIEW

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/FileId.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+package com.hedera.node.app.bbm.contracts;
+
+import com.google.common.collect.ComparisonChain;
+import com.hedera.hapi.node.base.FileID;
+import com.hedera.node.app.bbm.utils.Writer;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+// TODO: duplicated. Will be deleted after https://github.com/hashgraph/hedera-services/pull/11385 is merged
+record FileId(long shardNum, long realmNum, long fileNum) implements Comparable<FileId> {
```
Comment (vtronkov) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/FileId.java:25]: Added from https://github.com/hashgraph/hedera-services/pull/11385

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/FileStore.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.bbm.contracts;
+
+// TODO: duplicated. Will be deleted after https://github.com/hashgraph/hedera-services/pull/11385 is merged
+enum FileStore {
```
Comment (vtronkov) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/FileStore.java:20]: Added from https://github.com/hashgraph/hedera-services/pull/11385

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/HederaFile.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import com.hedera.node.app.service.mono.files.HFileMeta;
+import com.hedera.node.app.state.merkle.disk.OnDiskValue;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+
+/** Holds the content and the metadata for a single data file in the store */
+@SuppressWarnings("java:S6218") // "Equals/hashcode methods should be overridden in records containing array fields"
+// not using this with equals
+// TODO: duplicated. Will be deleted after https://github.com/hashgraph/hedera-services/pull/11385 is merged
+record HederaFile(
```
Comment (vtronkov) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/HederaFile.java:29]: Added from https://github.com/hashgraph/hedera-services/pull/11385

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/SystemFileType.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+ */
+
+package com.hedera.node.app.bbm.contracts;
+
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.Map;
+
+// TODO: duplicated. Will be deleted after https://github.com/hashgraph/hedera-services/pull/11385 is merged
+enum SystemFileType {
```
Comment (vtronkov) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/SystemFileType.java:24]: Added from https://github.com/hashgraph/hedera-services/pull/11385

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/ContractBytecodesDumpUtils.java [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+
+        final var deletedContractIds = deletedSmartContract.stream()
+                .map(c -> c.accountId().accountNum().intValue())
+                .toList();
+        return new Contracts(contractContents, deletedContractIds, smartContracts.size());
+    }
+
+    // TODO: duplicated. Will be deleted after https://github.com/hashgraph/hedera-services/pull/11385 is merged
+    @NonNull
+    private static Map<FileId, HederaFile> gatherModFiles(VirtualMap<OnDiskKey<FileID>, OnDiskValue<File>> source) {
```
Comment (vtronkov) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/ContractBytecodesDumpUtils.java:109]: Added from https://github.com/hashgraph/hedera-services/pull/11385

**Review by tinker-michaelj — CHANGES_REQUESTED** [review]
Looking very clean, noted one problem that mod-service stores bytecode in its own `VirtualMap`.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/ContractBytecodesDumpUtils.java [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+                .filter(HederaAccount::smartContract)
+                .toList();
+        final var deletedSmartContract =
+                smartContracts.stream().filter(HederaAccount::deleted).toList();
+
+        final var extractedFiles = gatherModFiles(files);
+
+        final var contractContents = new ArrayList<Contract>(ESTIMATED_NUMBER_OF_CONTRACTS);
+        for (final var smartContract : smartContracts) {
+            final var fileId = new FileID(0, 0, smartContract.accountId().accountNum()); // TODO: not sure
```
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/ContractBytecodesDumpUtils.java]: @tinker-michaelj  Fixed 👍 Can you review it now?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/ContractBytecodesDumpUtils.java [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+            @NonNull final Path path,
+            @NonNull final VirtualMap<OnDiskKey<AccountID>, OnDiskValue<Account>> accounts,
+            @NonNull final DumpCheckpoint checkpoint) {
+        final var dumpableAccounts = AccountDumpUtils.gatherAccounts(accounts, HederaAccount::fromMod);
+        final var contracts = getModContracts(dumpableAccounts);
+        final var sb = generateReport(contracts);
+        try (@NonNull final var writer = new Writer(path)) {
+            writer.writeln(sb.toString());
+            System.out.printf(
+                    "=== mono contract bytecodes report is %d bytes at checkpoint %s%n",
```
Comment (david-bakin-sl) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/ContractBytecodesDumpUtils.java:79]: Remove `mono` here and at L150 so mono and mod can compare equal

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/ContractBytecodesDumpUtils.java [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+                    writer.getSize(), checkpoint.name());
+        }
+    }
+
+    private static Contracts getModContracts(HederaAccount[] dumpableAccounts) {
```
Comment (david-bakin-sl) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/ContractBytecodesDumpUtils.java:84]: Consider moving to `ContractUtils.java` - renaming the current `getContracts` to `getMonoContracts`.

Also would be nice - but not required by any means! - if all or most of `ContractUtils.getAllContractContents` could be reused or adapted.

**Review by david-bakin-sl — APPROVED** [review]
LGTM with some optional items (but also pending MichaelT's check that getting the contract bytecode store is correct).

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/state/InitialModServiceContractSchema.java [hunk 9]
```diff
@@ -54,7 +54,7 @@ public class InitialModServiceContractSchema extends Schema {
     private static final Logger log = LogManager.getLogger(InitialModServiceContractSchema.class);
     public static final String STORAGE_KEY = "STORAGE";
     public static final String BYTECODE_KEY = "BYTECODE";
-    private static final int MAX_BYTECODES = 50_000_000;
+    public static final int MAX_BYTECODES = 50_000_000;
```
[the remark this answers is not in the document]
Comment (vtronkov) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/state/InitialModServiceContractSchema.java:57]: Yes...I was using it but then I deleted the code. Reverted 👍

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/ContractBytecodesDumpUtils.java (hunk 7 above)
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/ContractBytecodesDumpUtils.java:79]: Removed in both places but do we compare the console print as well? If so we should remove these **printf** statements in all the dumper

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/ContractBytecodesDumpUtils.java (hunk 8 above)
Comment (vtronkov) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/ContractBytecodesDumpUtils.java]: Okay, I've moved the method inside `ContractUtils` 👍 
I didn't get the second part. What do you mean by reusing or adapting the `ContractUtils.getAllContractContents`? Are we using the same method somewhere else?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/ContractBytecodesDumpUtils.java (hunk 7 above)
Comment (david-bakin-sl) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/ContractBytecodesDumpUtils.java]: ah.  didn't notice this didn't go into the csv, duh! No, don't worry about it.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/ContractBytecodesDumpUtils.java (hunk 8 above)
Comment (david-bakin-sl) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/ContractBytecodesDumpUtils.java]: looks like a bunch of very similar code ... probably _not_ worth changing at this stage.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/ContractUtils.java [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+    }
+
+    private static int getContractId(HederaAccount contract) {
+        if (contract.accountId() == null || contract.accountId().accountNum() == null) {
+            return 0;
+        }
+        return contract.accountId().accountNum().intValue();
+    }
+
+    private static ReadableKVState<ContractID, Bytecode> getContractStore() {
```
Comment (tinker-michaelj) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/ContractUtils.java:161]: ⚠️ Unless I'm missing some magic here, isn't this just constructing an empty `KVState` and returning it?

Which means the mod service bytecodes will always appear to be missing?

I would have expected `dumpModContractBytecodes()` to receive e.g. a `VirtualMap<OnDiskKey<ContractID>, OnDiskValue<Bytecode>>` which was got [here](https://github.com/hashgraph/hedera-services/blob/dfc70e33db0bf73be47a51840e60bee686802061/hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/StateDumper.java#L115) via a call to 

```
state.getChild(
  state.findNodeIndex(ContractService.NAME,
  InitialModServiceContractSchema.BYTECODE_KEY)
```

**Review by tinker-michaelj — APPROVED** [review]
LGTM, tyvm @vtronkov !

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/ContractUtils.java (hunk 10 above)
Comment (david-bakin-sl) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/bbm/contracts/ContractUtils.java]: Is your last comment still valid?  I see some handling of deleted contracts by checking the accounts for the deleted flag.

**Review by david-bakin-sl — APPROVED** [review]
lgtm

**Review by tinker-michaelj — APPROVED** [review]
LGTM!

**Pull request merged**